sycl/doc/extensions/KernelProperties/KernelProperties.asciidoc

```diff
@@ -0,0 +1,519 @@
+                               PropertyList properties,
+                               const WorkgroupFunctionType &kernelFunc);
+
+  template <typename KernelName, typename WorkgroupFunctionType, int dimensions, typename PropertyList>
+  void parallel_for_work_group(range<dimensions> numWorkGroups,
+                               range<dimensions> workGroupSize,
+                               PropertyList properties,
+                               const WorkgroupFunctionType &kernelFunc);
+}
+}
+```
+
+Passing properties as an argument in this way allows properties to be
+associated with a kernel function without modifying its type. This enables
+the same kernel function (e.g. a lambda) to be submitted multiple times with
+different properties, or for libraries building on SYCL to add properties
+(e.g. for performance reasons) to user-provided kernel functions.
+
+All the properties defined in this extension have compile-time values. However,
+an implementation may support additional properties which could have run-time
+values. When this occurs, the `properties` parameter may be a `property_list`
+containing a mix of both run-time and compile-time values, and a SYCL
+implementation should respect both run-time and compile-time information when
+determining the correct way to launch a kernel. However, only compile-time
+information can modify the compilation of the kernel function itself.
+
+A simple example of using this extension to set a required work-group size
+and required sub-group size is given below:
+
+```c++
+sycl::ext::oneapi::property_list properties{sycl::ext::oneapi::work_group_size_v<8, 8>,
+                                            sycl::ext::oneapi::sub_group_size_v<8>};
+q.parallel_for(range<2>{16, 16}, properties, [=](id<2> i) {
+  a[i] = b[i] + c[i];
+}).wait();
+```
+
+== Encoding Properties into a Kernel
+
+In other situations it may be useful to encode a kernel's properties directly
+into its type, to ensure that a kernel cannot be launched without a property
+that it depends upon for correctness.
+
+To enable this use-case, this extension adds a mechanism for implementations to
+extract a property list from a kernel functor, if a kernel functor declares
+a `property_list` member variable. Note that this member variable must be
+`static constexpr`, and kernel functors can therefore only encode properties
+with compile-time values.
+
+The example below shows how the kernel from the previous section could be
+rewritten to leverage an embedded property list:
+
+```c++
+struct KernelFunctor {
+
+  KernelFunctor(sycl::accessor<int, 2> a,
+                sycl::accessor<int, 2> b,
+                sycl::accessor<int, 2> c) : a(a), b(b), c(c)
+  {}
+
+  void operator()(id<2> i) const {
+    a[i] = b[i] + c[i];
+  }
+
+  static constexpr auto properties =
+    sycl::ext::oneapi::property_list{sycl::ext::oneapi::work_group_size_v<8, 8>,
+                                     sycl::ext::oneapi::sub_group_size_v<8>};
+
```

Comment:
The `_v` looks like an implementation detail.
```suggestion
  static constexpr auto properties =
    sycl::ext::oneapi::property_list{sycl::ext::oneapi::work_group_size<8, 8>,
                                     sycl::ext::oneapi::sub_group_size<8>};
```
While it does appear in the current SYCL spec by oposition to the SYCL 1.2.1 device selector for example, this is probably not necessary here because we are on new territories.
We could keep private the types themselves and just expose the variables like above.